types/ember-data/index.d.ts

```diff
@@ -25,7 +25,30 @@ import AdapterRegistry from 'ember-data/types/registries/adapter';
 type ModelKeys<Model extends DS.Model> = Exclude<keyof Model, keyof DS.Model>;
 
 type AttributesFor<Model extends DS.Model> = ModelKeys<Model>; // TODO: filter to attr properties only (TS 2.8)
-type RelationshipsFor<Model extends DS.Model> = ModelKeys<Model>; // TODO: filter to hasMany/belongsTo properties only (TS 2.8)
+
+type UnboxBelongsTo<T> =
+    T extends DS.AsyncBelongsTo<infer C> ? C
+        : T extends DS.Model ? T
+        : T extends DS.AsyncHasMany<any> ? never
+        : T extends DS.SyncHasMany<any> ? never
+        : DS.Model;
+
+type UnboxHasMany<T> =
+    T extends DS.AsyncHasMany<infer C> ? C
+      : T extends DS.SyncHasMany<infer C> ? C
+      : T extends DS.AsyncBelongsTo<any> ? never
+      : T extends DS.Model ? never
+      : DS.Model;
```

Comment:
I haven't had a chance to run this locally, but the `DS.Model` fallback in the final condition here makes me wonder whether `BelongsToModels<Model>` will turn something like
```ts
interface Model {
   nonPersistedField1: string;
   nonPersistedField2: string;
   nonPersistedField3: string;
   relatedThings: DS.AsyncHasMany<Foo>
}
```
into 
```ts
interface Model {
   nonPersistedField1: DS.Model;
   nonPersistedField2: DS.Model;
   nonPersistedField3: DS.Model;
   relatedThings: Foo
}
```
Can we introduce some more test cases that make it clear that this utility type works the way you want it to?

I'm particularly interested in cases like

* `DS.attr()` (no property type)
* a regular (not `DS.attr` )string class field that's present on an e-d model class)
* Models with no relationships to other models at all
* Models with computed properties

some of the patterns above are inadvisable, but they are supported and exist in the real world. We should make sure that they don't _break_ with this change.

Reply:
That's how it works currently, and this PR doesn't change this behavior, as confirmed by a test case( [before](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51852/commits/9403bc13b7184ffaf3b05c34b4dae74365dfdaa3#diff-6fd426c31e33023aa19e22cae72a4f94d106a4d3575ebffe9f77260406ee6f15R116), [after](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51852/files#diff-6fd426c31e33023aa19e22cae72a4f94d106a4d3575ebffe9f77260406ee6f15R115) ) written by your suggestion.

I'm open to also fix this issue, and start respecting `DS.PromiseObject`, `DS.PromiseManyArray`, and the rest from the  [my previous attempt](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42839/files#diff-f84a1c4948e4f7317ca9d56ad8a88c9ae199d5fb9cb327ae8d1add268f2c725dR30-R39). Please let me know if I should do it?

cc @jamescdavis